File: Daily Resources/Day11/DSA - Tree Solutions/tree_matching.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 2e5 + 5;
vector<int> tree[N];
int dp[N][2];

void dfs(int u, int parent) {
    for (int v : tree[u]) {
        if (v == parent) continue;
        dfs(v, u);
        dp[u][0] += max(dp[v][0], dp[v][1]);
    }
    for (int v : tree[u]) {
        if (v == parent) continue;
        dp[u][1] = max(dp[u][1], dp[u][0] - max(dp[v][0], dp[v][1]) + dp[v][0] + 1);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n;
    cin >> n;
    for (int i = 0, a, b; i < n - 1; ++i) {
        cin >> a >> b;
        tree[a].push_back(b);
        tree[b].push_back(a);
    }

    dfs(1, 0);
    cout << max(dp[1][0], dp[1][1]) << '\n';
}

File: Daily Resources/Day11/DSA - Tree Solutions/tree_matching.py
import sys
sys.setrecursionlimit(2 * 10**5 + 10)

n = int(input())
tree = [[] for _ in range(n + 1)]

for _ in range(n - 1):
    a, b = map(int, input().split())
    tree[a].append(b)
    tree[b].append(a)

dp = [[0, 0] for _ in range(n + 1)]  # dp[u][0] = not matched, dp[u][1] = matched

def dfs(u, parent):
    for v in tree[u]:
        if v == parent:
            continue
        dfs(v, u)
        dp[u][0] += max(dp[v][0], dp[v][1])
    
    for v in tree[u]:
        if v == parent:
            continue
        dp[u][1] = max(dp[u][1], dp[u][0] - max(dp[v][0], dp[v][1]) + dp[v][0] + 1)

dfs(1, 0)
print(max(dp[1][0], dp[1][1]))
